core/solver/bicg.cpp

```diff
@@ -0,0 +1,196 @@
+
+    if (transposable_system_matrix) {
+        trans_A = transposable_system_matrix->transpose();
+    } else {
+        csr_system_matrix_unique_ptr = copy_and_convert_to<CsrMatrix>(
+            system_matrix_->get_executor(),
+            const_cast<LinOp *>(system_matrix_.get()));
+
+        csr_system_matrix_unique_ptr->set_strategy(
+            std::make_shared<typename CsrMatrix::classical>());
+
+        auto csr_system_matrix_ = csr_system_matrix_unique_ptr.get();
+
+        trans_A = csr_system_matrix_->transpose();
+    }
+
+    auto trans_preconditioner_tmp =
+        as<const Transposable>(get_preconditioner().get());
+    auto trans_preconditioner = trans_preconditioner_tmp->transpose();
+
+    system_matrix_->apply(neg_one_op.get(), dense_x, one_op.get(), r.get());
+    // r = r - Ax =  -1.0 * A*dense_x + 1.0*r
+    r2->copy_from(r.get());
+    // r2 = r
+    auto stop_criterion = stop_criterion_factory_->generate(
+        system_matrix_, std::shared_ptr<const LinOp>(b, [](const LinOp *) {}),
+        x, r.get());
+
+    int iter = -1;
+
+    while (true) {
+        get_preconditioner()->apply(r.get(), z.get());
+        trans_preconditioner->apply(r2.get(), z2.get());
+        z->compute_dot(r2.get(), rho.get());
+
+        ++iter;
+        this->template log<log::Logger::iteration_complete>(this, iter, r.get(),
+                                                            dense_x);
+        if (stop_criterion->update()
+                .num_iterations(iter)
+                .residual(r.get())
+                .solution(dense_x)
+                .check(RelativeStoppingId, true, &stop_status, &one_changed)) {
+            break;
+        }
+
+        exec->run(bicg::make_step_1(p.get(), z.get(), p2.get(), z2.get(),
+                                    rho.get(), prev_rho.get(), &stop_status));
+        // tmp = rho / prev_rho
+        // p = z + tmp * p
+        // p2 = z2 + tmp * p2
+        system_matrix_->apply(p.get(), q.get());
+        trans_A->apply(p2.get(), q2.get());
+        p2->compute_dot(q.get(), beta.get());
+        exec->run(bicg::make_step_2(dense_x, r.get(), r2.get(), p.get(),
+                                    q.get(), q2.get(), beta.get(), rho.get(),
+                                    &stop_status));
+        // tmp = rho / beta
+        // x = x + tmp * p
+        // r = r - tmp * q
+        // r2 = r2 - tmp * q2
+        swap(prev_rho, rho);
+    }
+}
+
+
+template <typename ValueType, typename IndexType>
+void Bicg<ValueType, IndexType>::apply_impl(const LinOp *alpha, const LinOp *b,
+                                            const LinOp *beta, LinOp *x) const
+{
+    auto dense_x = as<matrix::Dense<ValueType>>(x);
+
+    auto x_clone = dense_x->clone();
+    this->apply(b, x_clone.get());
+    dense_x->scale(beta);
+    dense_x->add_scaled(alpha, x_clone.get());
+}
+
+
+#define GKO_DECLARE_BICG(_type) class Bicg<_type>
+GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_BICG);
```

Comment:
If you want to use the `IndexType`, you would actually need to use the following macro instead `GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE`.